### ISSUE TITLE
Trying support for PC-Speaker, Covox, and Gravis UltraSound

### ISSUE THREAD

**Issue by 4nd3r50ncr** [crazii/SBEMU#43]
What do you think about us evolving this business into gravis ultrasound “GUS”, and Covox, and PC-Speaker? Oh, I was forgetting, we didn't need to port the highlight from sbemu-x so that when the connection is successful, the letters turn green, otherwise they turn red.

**Comment by Torinde** [comment]
Yes, that would be good!

AWE32/64 would also be interesting - maybe reusing code from PCem/86Box/VMusic or SoundFX (related: dosbox-staging/dosbox-staging/issues/1682) or ALSA/Linux drivers.

Although probably first should come the switch to the HX port-trapping API (maybe requiring also to upstream [Get Interrupt Context](https://github.com/Baron-von-Riedesel/HX/issues/11#issuecomment-1818276436)) https://github.com/crazii/SBEMU/issues/40 sbemu-x/sbemu-x/issues/12

**Comment by crazii** [comment]
Yeah people in VOGONS mentioned those, and MIDI too. I don't refuse those at all. I think all of those will be done eventually. Are you planning to start (one of them) right now? I think it'll helps a lot if you do.
For myself, I would pick pc speaker or MIDI first because they looks more needed (probably, for most users. not really sure), but you can pick any one that you're interested in, if you're planning to.

**Comment by volkertb** [comment]
@crazii Wouldn't it be better to go for a modular emulation core that would focus on the key functionality, namely the I/O port trapping and ISA DMA emulation part, with support in both real mode and protected mode games? And then provide an API for loadable modules that people could work on in separate projects?

Something I described [here](https://github.com/sbemu-x/sbemu-x/issues/5#issuecomment-1773304621).

That way, 100% focus could be given on getting the port trapping stuff to work and work well, with various other projects focusing on developing the best possible modular backends, both complete software emulation cores for specific hardware (Sound Blaster Pro, GUS, etc) and simple "redirection" drivers in the vein of [ADLiPT](https://github.com/pdewacht/adlipt) and [SoftMPU](https://github.com/bjt42/softmpu). Heck, both those projects could then piggyback on this one.

**Comment by crazii** [comment]
Yes that's another way to get the midi (at least SoftMPU) to work.
I bet the QPI emu is the breach: we just add HDPMI32i dependency to QPIEMU so that it adds the same pot trapping for PM mode, and everything is done. The dependency doesn't hurt because it's a plugin, not jemm itself. BUT I'm not sure if it's doable. and it also has more RM-PM mode transitions.

**Comment by 4nd3r50ncr** [comment]
@crazii you forgot to add me as one of the collaborators in the project, I was already preparing myself with my time to help create new functions, I need you to give me your compilation instructions with RHIDE directly through DOS, creating a grp file of the instructions and the conversion with gpr2mak, in a complete way, so that I can do it more directly, then these instructions will be passed to the Linux compilation makefile. I need your instructions or your pre-configured zipped compilation package, it may be right here for me.

**Comment by crazii** [comment]
I don't know how to add collaborators. Let me check.
I've added a makefile for dos `makefile.dos` you can perform build & run on the fly under dos now. `make -f makefile.dos` `make -f makefile.dos clean` Let me know if the grp & grp2mak are still needed.

**Comment by crazii** [comment]
I added @thp @volkertb @4nd3r50ncr as collaborators, but I don't know the difference, Doesn't creating PR work for you?

**Comment by crazii** [comment]
Do you mean you need a zip pack for DJGPP tools? you can download it from here https://www.delorie.com/djgpp/zip-picker.html
you need multiple zips and extract them into a same location.

**Comment by crazii** [comment]
I updated dos djgpp setup steps in readme https://github.com/crazii/SBEMU?tab=readme-ov-file#installing-djgpp-on-dos

**Comment by 4nd3r50ncr** [comment]
@crazii You removed me again as one of the collaborators, I compressed the image of my test table system, I wanted you to test it, burn the image and run it on a real machine, can you try it in emulation? It wouldn't be ideal, I need you to test it on real hardware. See how it will work on your modern computer. It is a copy of a 2GB pendrive image. Type VER in commander.

https://drive.google.com/file/d/1jwcDUO5yPZ3CZICWtTdcUw3IRQfb9wWA/view?usp=drive_link

**Comment by crazii** [comment]
No I didn't remove you, as I can see you in the settings. I'm totally inaccessible to any google sites (including youtube) from my country/region.
You may ask help for others :)

**Comment by 4nd3r50ncr** [comment]
I planned to, but I'm not part of the group of developers, I was even programming myself for this, and organizing the compilation environment to stay focused, I was doing tests and everything, checking bugs. I had sent the invitation but then they removed it from the development participation team. It gave me discouragement.

**Comment by 4nd3r50ncr** [comment]
I planned to, but I'm not part of the group of developers, I was even programming myself for this, and organizing the compilation environment to stay focused, I was doing tests and everything, checking bugs. I had sent the invitation but then they removed it from the development participation team. It gave me discouragement.

**Comment by volkertb** [comment]
@4nd3r50ncr Is there a way we could rekindle your encouragement? Because with an ambitious project like this, the more people that contribute, the better. @crazii has been doing all the heavy lifting w.r.t. work on the actual source code, with some help from @thp. So far, all I have been able to contribute is CI/CD automation stuff.

Could you maybe create a Pull Request and have us review your code?

That's basically all that's needed for you to effectively become a contributor, as far as I know.

**Comment by volkertb** [comment]
@4nd3r50ncr [As you can see](https://github.com/crazii/SBEMU/graphs/contributors), committing code effectively makes you a "contributor".

So just open a PR. You can start with a "draft" PR, if you think it needs more work before it's ready. It's always good to have others give feedback, even as you work on it. :slightly_smiling_face:

**Comment by 4nd3r50ncr** [comment]
> What do you think about us evolving this business into gravis ultrasound “GUS”, and Covox, and PC-Speaker? Oh, I was forgetting, we didn't need to port the highlight from sbemu-x so that when the connection is successful, the letters turn green, otherwise they turn red.

Right, as far as I know, dosbox has these features, so it would be easier to port them than to create them from scratch. I'm going to take the day today to analyze the dosbox code to look for these features. And we continue as much as possible, and depending on time, I work every day in the early hours of the morning, so if there is time for me during the day I will continue. Now we just need to be patient 😅. Hugs friends!

**Comment by 4nd3r50ncr** [comment]
> What do you think about us evolving this business into gravis ultrasound “GUS”, and Covox, and PC-Speaker? Oh, I was forgetting, we didn't need to port the highlight from sbemu-x so that when the connection is successful, the letters turn green, otherwise they turn red.

Right, as far as I know, dosbox has these features, so it would be easier to port them than to create them from scratch. I'm going to take the day today to analyze the dosbox code to look for these features. And we continue as much as possible, and depending on time, I work every day in the early hours of the morning, so if there is time for me during the day I will continue. Now we just need to be patient 😅. Hugs friends!

**Comment by 4nd3r50ncr** [comment]
> What do you think about us evolving this business into gravis ultrasound “GUS”, and Covox, and PC-Speaker? Oh, I was forgetting, we didn't need to port the highlight from sbemu-x so that when the connection is successful, the letters turn green, otherwise they turn red.

Right, as far as I know, dosbox has these features, so it would be easier to port them than to create them from scratch. I'm going to take the day today to analyze the dosbox code to look for these features. And we continue as much as possible, and depending on time, I work every day in the early hours of the morning, so if there is time for me during the day I will continue. Now we just need to be patient 😅.

**Comment by crazii** [comment]
> These are the three files responsible for the hardware. So come on, Covox is [Disney](https://github.com/joncampbell123/dosbox-x/blob/master/src/hardware/disney.cpp), then [GUS ](https://github.com/joncampbell123/dosbox-x/blob/master/src/hardware/gus.cpp)(Gravis UltraSound), and [PC-Speaker](https://github.com/joncampbell123/dosbox-x/blob/master/src/hardware/pcspeaker.cpp). These are the codes responsible for the hardware. Right @crazii ?

Yes, they are.
And here's some programming guides I found:
https://wiki.preterhuman.net/Programmer%27s_Guide_to_the_Disney_Sound_Source
https://wiki.osdev.org/PC_Speaker
https://archive.gamedev.net/archive/reference/articles/article448.html

I wonder if they actually help but maybe use them as references for better understanding DOSBox codes.

**Comment by hjnijlunsing** [comment]
For the Realtek HDA interface, would enabling the PC Beep hidden register suffice gor PC Speaker support? (https://docs.kernel.org/sound/hd-audio/realtek-pc-beep.html)

Current behaviour on Realtek HDA:
C:\> echo ^G
<Beeps>
SBEMU
C:\> echo ^G
<Silent>

**Comment by volkertb** [comment]
If I may share my opinion:

I think we should first focus on improving the stability and compatibility of the current Sound Blaster, Adlib and MPU-401 emulation support. Those are, after all, the most important DOS-era sound devices in terms of game support.

The Covox Speech Thing and Disney Sound Source are interesting as actual physical devices to play with (especially on modern PCs), but adding support for emulating those two devices in software has very limited added value, since practically every game that supports those devices also support Sound Blaster.

PC Speaker emulation support would perhaps be a useful one to add, assuming that modern systems no longer support it natively, which I don't believe is the case, at least for motherboards and laptops that still have a CSM for legacy BIOS support.

But even then, it would be better to split that off into a separate issue.

And as for Gravis Ultrasound support, I have to say, even though I had a GUS in my teenage computing years, and was extremely loyal to that standard, the only real advantage of supporting it would be for sound support in demos. In terms of game support, there are very few games that offer an improve sound experience on the GUS compared to Sound Blaster 16 combined with General MIDI or MT-32 over MPU-401.  So it would be cool to add it at some point, but I wouldn't prioritize it. 

Ultimately, as requests for additional devices (both hardware backends and emulated devices) increase, I think it may be prudent to consider evolving SBEMU to a modular emulation framework, with separate loadable modules (perhaps DLLs) for different emulators and backends.

@crazii Curious about your thoughts on this.

**Comment by crazii** [comment]
Introducing loadable modules will increase the complexity of the code, first DOS DLLs are rare and not essential and merely no advantages (DLLs are shared among processes/exes to save memory/disk space, it also forces modularizing the source code in a way but still not necessary), second, you need to resolve conflicts/dependencies among different modules, i.e. module A want trapping an io port which is already trapped by module B, which is likely to happen with current MIDI through UART by @jiyunomegami  and pure soft MPU in the future - Yes they will conflict without modules anyways but can be solved straight forward / hard coded. 

I do agree that we need make it stable first, but if someone want to add more features without refactoring current code which will make it more unstable. I think it's OK.

**Comment by andersrodrig** [comment]
[sbemu.zip](https://github.com/user-attachments/files/31350991/sbemu.zip)
https://github.com/andersrodrig/SBEMU

@crazii @volkertb @thp @Torinde @Baron-von-Riedesel Is there still any interest in this implementation GUS?

EDIT: Well, I implemented a few things based on DOSBox; it's 90% complete though, of course, I couldn't do it all on my own. But there's some small detail preventing it from working. If you want to see how it turned out, here it is just compile it on Linux. GUS is implemented, but I’ve tried every possible way to get it working port outputs, etc... without success.

EDIT: As promised, GUS, COVOX, PC-SPEAKER. Compiling on Windows is easy; if DJGPP is on the C: drive, simply run build.bat and it will compile. What works and has been tested? The Sound Blaster is intact—nothing has been altered. I implemented other features based on DOSBox—such as PC speaker, Covox (or Disney Sound Card, whatever you prefer to call it), and Gravis UltraSound (GUS). However, regarding the GUS: it is present and working in the build, and the code isn't causing any corruption, but games can't seem to detect the card. I’ve tried everything; could it be a specific configuration setting I'm missing? I’ve never actually owned a GUS!

**Comment by andersrodrig** [comment]
Close issue.
